### PR TITLE
impl new subspace naming strategy

### PR DESCRIPTION
**Pull request by sadlerap**
This should prevent subspace names from growing longer from the level of subspace nesting.

### CODE REVIEW

File: pkg/space/space.go
Comment (alexeykazakov) [inline pkg/space/space.go:68]: Let's also check that the parent space name is actually longer than 6 characters. Just to be 100% on the safe side.

File: pkg/space/space.go
Comment (alexeykazakov) [inline pkg/space/space.go:73]:
```suggestion
	return fmt.Sprintf("%v-%v", parentSpaceName, string(spacerequest.UID[:5])) // we only get the first 5 chars from UID so we can keep the name within the length limits
```

**Review by alexeykazakov — COMMENTED** [review]
Looks good! Just a couple of minor comments.

File: pkg/space/space.go
Comment (sadlerap) [reply pkg/space/space.go]: Fixed in 46b3f1e.

File: pkg/space/space.go
Comment (sadlerap) [reply pkg/space/space.go:74]: Fixed in 46b3f1e.

**Comment by sadlerap** [comment]
Hmmm, looks like uploading coverage info to codecov.io is [failing](https://github.com/codeready-toolchain/host-operator/actions/runs/6550083733/job/17788529292).  Not sure what's going on, since it looks like we're getting a 404, but [their status page](https://status.codecov.com/) is all green.

**Comment by alexeykazakov** [comment]
Codecove upload fails pretty often unfortunately. Re-running the job usually helps.

File: pkg/space/space.go
Comment (mfrancisc) [inline pkg/space/space.go:69]: minor thing - maybe adding a small example to make it clear to the reader that we are removing the UID suffix from the parent space name before using it for the sub space name creation. WDYT ? 
```suggestion
		// take of 6 characters to include the dash
		// for example if the parentSpaceName is "parentSpace-01234" then we keep only "parenSpace", removing the 5char UID suffix from the end. 
```

**Review by mfrancisc — APPROVED** [review]
Nice 👍 . Thanks for improving this 🙏 

I have only one minor comment.

File: pkg/space/space.go
Comment (sadlerap) [reply pkg/space/space.go]: sounds good to me - I've adjusted some of the comments with a7f983f, hopefully things are a bit clearer.

Comment (mfrancisc) [reply pkg/space/space.go]: looks good ! thanks !

File: pkg/space/space.go
Comment (MatousJobanek) [inline pkg/space/space.go:77]: how do you want to ensure that the already existing sub-spaces will be re-used after this change is merged?
1. there is a SpaceRequest CR in RHTAP instance
2. for the SpaceRequest CR, there is a sub-space `johny123` created
3. you merge the PR
4. the reconciler is kicked and it goes to this section https://github.com/codeready-toolchain/host-operator/blob/932b437277941dce8cc27b1e319931fd980d3cb0/controllers/spacerequest/spacerequest_controller.go#L183-L187
5. it uses this new format and expects `johny-123`, it doesn't find any, so it creates a new sub-space

the same would happen for the spaces whose parent has a name longer than 6

File: pkg/space/space.go
Comment (MatousJobanek) [inline pkg/space/space.go:68]: out of curiosity, why did you decide to use only 6 first characters?

**Review by MatousJobanek — CHANGES_REQUESTED** [review]
Hi, thanks a lot for your PR. :+1: :1st_place_medal: 
I think that the code doesn't cover the case with already existing sub-spaces. Could you please change that and also cover such use-case in unit-tests as well as in e2e tests?
Let me know if I missed some important detail .

File: pkg/space/space_test.go
Comment (MatousJobanek) [inline pkg/space/space_test.go:67]: Let's not use the same SpaceRequest for both SubSpaces to make sure that the subSubSpace doesn't have the same name as its parent (subSpace)

File: pkg/space/space_test.go
Comment (MatousJobanek) [inline pkg/space/space_test.go:70]: based on the changes from the previous comment, let's verify that it's using UID from the right SpaceRequest
```suggestion
	expectedSubSubSpace := spacetest.NewSpace(test.HostOperatorNs, fmt.Sprintf("parentspace-%s", sr.UID[:5]),
```

File: pkg/space/space_test.go
Comment (MatousJobanek) [inline pkg/space/space_test.go:81]: and also that the names are different 
```suggestion
	assert.Equal(t, len(subSpace.Name), len(subSubSpace.Name))
	assert.NotEqual(t, subSpace.Name, subSubSpace.Name)
```

File: pkg/space/space.go
Comment (mfrancisc) [reply pkg/space/space.go:77]: good point ! But do we have any existent subSubSpace ? 

I mean the `parentSpaceName` will be shortened only if it's a subSpace it self, not only if the name is longer than 6:
`if parentSpace.Spec.ParentSpace != "" && len(parentSpaceName) > 6`

I thought right now we are using this feature only for 1 level of inheritance : `parentSpace --> subSpace` , and not yet `parentSpace --> subSpace --> subSubSpace` but maybe I'm wrong.

Comment (sadlerap) [reply pkg/space/space.go:77]: In this particular case, could we list with a label selector for the correct parent space?
```go
subSpaceList := &toolchainv1alpha1.SpaceList{}
err := r.Client.List(ctx, subSpaceList,
    client.InNamespace(r.Namespace),
    client.MatchingLabels{
        toolchainv1alpha1.SpaceRequestLabelKey:          spaceRequest.GetName(),
        toolchainv1alpha1.SpaceRequestNamespaceLabelKey: spaceRequest.GetNamespace(),
        toolchainv1alpha1.ParentSpaceLabelKey:           parentSpace.GetName(),
    })
```

We'd have to handle the case where we find more than one matching space, but I can add a test for that too.

File: pkg/space/space_test.go
Comment (sadlerap) [reply pkg/space/space_test.go]: Fixed in 35eb34a.

File: pkg/space/space_test.go
Comment (sadlerap) [reply pkg/space/space_test.go]: Fixed in 35eb34a.

File: pkg/space/space_test.go
Comment (sadlerap) [reply pkg/space/space_test.go:85]: Fixed in 35eb34a.

**Comment by sadlerap** [comment]
I'll look into writing some e2e tests for this.  If we have a subspace, we can't guarantee that it follows the new naming convention established here, so I imagine most places where we try to get a subspace by its name will need to turn into a list call with a label selector for its parent.

File: pkg/space/space.go
Comment (MatousJobanek) [reply pkg/space/space.go:77]: > good point ! But do we have any existent subSubSpace ?

not of subSubSpace, but of subSpace yes. There is one more change in the logic of generating the name apart from the `if` statement that you mentioned above - originally, we didn't add the name between the prefix (parent name) and suffix (UID of the SpaceRequest CR). so the result was for example `parentspace1a2b3cd`. For better readability, we would like to add a hyphen between the prefix and suffix so its' `parentspace-1a2b3cd`, which however brings the challenge of keeping the backward compatibility.

The list for checking the existence should work. We had it there even before but we used it in a combination with generating the name nondeterministically. We modified it to the `Get` (as it is now) because we saw that the informer cache of the client wasn't sometimes updated quickly enough so the `List` returned an empty list even when one Space was already created. However, since we didn't use deterministic approach to generating the names, the we sometimes ended up with multiple Spaces per SpaceRequest CR.
Using the `List` should be fine, because it uses a fixed format for the names, so if the cache is not updated with the latest Space CR, then the next attempt to create the Space CR should fail with conflict, which totally OK. As a benefit, we don't have to rely on the naming format, but just on the labels set at Space CR.

**Comment by MatousJobanek** [comment]
As for the e2e tests, take a look at the migration tests https://github.com/codeready-toolchain/toolchain-e2e/tree/master/test/migration We don't have a case for sub-spaces there yet, so if you add one, then it should cover exactly what we need. Just follow the format that is used there.

File: pkg/space/space.go
Comment (mfrancisc) [reply pkg/space/space.go:77]: > There is one more change in the logic

My bad ! I missed that we are also adding the `-` between parentName and suffix 🤦‍♂️ . I don't remember why we haven't added that in the first place.

And indeed, if we want to change the naming convention of subSpaces as well, we could use `List` by labels while keeping the name deterministic.

Comment (sadlerap) [reply pkg/space/space.go:77]: I've replaced most of the direct fetches with list calls in 563be3d (and added some tests in ff9eec2).  As far as I can tell, the only place that uses `SubSpaceName` anymore is some tests, and I'm half-tempted to refactor those tests to make those calls indirect, though I'll probably punt that off for a future PR if I do.

File: pkg/space/space_test.go
Comment (mfrancisc) [inline pkg/space/space_test.go:87]: We should rename sr and expect that as name and namespace in the labels:
```suggestion
func TestNewSubSubSpace(t *testing.T) {
	// given
	srClusterRoles := []string{commoncluster.RoleLabel(commoncluster.Tenant)}
	sr := spacerequesttest.NewSpaceRequest("jane", "jane-tenant",
		spacerequesttest.WithTierName("appstudio"),
		spacerequesttest.WithTargetClusterRoles(srClusterRoles))
	parentSpace := spacetest.NewSpace(test.HostOperatorNs, "parentSpace")
	subSpace := NewSubSpace(sr, parentSpace)
	// create another spacerequest in the subSpace namespace
	sr2 := spacerequesttest.NewSpaceRequest("jane2", subSpace.GetName()+"-tenant",
		spacerequesttest.WithTierName("appstudio"),
		spacerequesttest.WithTargetClusterRoles(srClusterRoles))

	// when
	subSubSpace := NewSubSpace(sr2, subSpace)

	// then
	expectedSubSubSpace := spacetest.NewSpace(test.HostOperatorNs, fmt.Sprintf("%s-%s", parentSpace.Name, sr2.UID[:5]),
		spacetest.WithSpecParentSpace(subSpace.GetName()),
		spacetest.WithTierName("appstudio"),
		spacetest.WithSpecTargetClusterRoles([]string{cluster.RoleLabel(cluster.Tenant)}),
		spacetest.WithLabel(toolchainv1alpha1.SpaceRequestLabelKey, sr2.GetName()),
		spacetest.WithLabel(toolchainv1alpha1.SpaceRequestNamespaceLabelKey, sr2.GetNamespace()),
		spacetest.WithLabel(toolchainv1alpha1.ParentSpaceLabelKey, subSpace.GetName()),
	)
	assert.Equal(t, expectedSubSubSpace, subSubSpace)

	// also assert that names don't grow in length as we increase nesting
	assert.Equal(t, len(subSpace.Name), len(subSubSpace.Name))
	assert.NotEqual(t, subSpace.Name, subSubSpace.Name)
}
```

**Review by mfrancisc — COMMENTED** [review]
few minor things

File: controllers/spacerequest/spacerequest_controller.go
Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller.go:211]: I like it - I'll see if I can come up with something more satisfying.

File: pkg/space/space_test.go
Comment (sadlerap) [reply pkg/space/space_test.go:87]: Fixed in bec33a7.

File: controllers/spacerequest/spacerequest_controller_test.go
Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller_test.go:1123]: I took this test out in eeb9e65, since it wasn't doing anything for coverage.

File: controllers/spacerequest/spacerequest_controller.go
Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller.go]: I've adjusted this in eeb9e65.

File: controllers/spacerequest/spacerequest_controller.go
Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller.go]: Fixed in eeb9e65.

File: controllers/spacerequest/spacerequest_controller.go
Comment (mfrancisc) [inline controllers/spacerequest/spacerequest_controller.go:169]: minor thing feel free to leave it as is - but how about `listSubSpace`? Since it's actually listing and not getting.

**Review by mfrancisc — APPROVED** [review]
Nice job 👍 

Thanks for addressing my comments!

**Comment by sadlerap** [comment]
/hold

I'd like https://github.com/codeready-toolchain/toolchain-e2e/pull/826 to merge before we merge this - the tests there make sure that this PR doesn't break anything.

Local testing seems to indicate things work, but I'll have higher confidence once it lands.

**Review by MatousJobanek — APPROVED** [review]
Looks good :+1: :rocket: Thanks a lot for the additional changes :1st_place_medal: :pray: 
I have just one comment regarding one unit test

**Comment by MatousJobanek** [comment]
> I'd like https://github.com/codeready-toolchain/toolchain-e2e/pull/826 to merge before we merge this - the tests there make sure that this PR doesn't break anything.
Local testing seems to indicate things work, but I'll have higher confidence once it lands.

It's always the best practice to create both PRs from branches that have the same name so our e2e logic can pair them and run e2e tests from the e2e PR (instead of master). In this way, you can verify that the changes in e2e tests work as expected with the changes from the operator.

File: controllers/spacerequest/spacerequest_controller_test.go
Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller_test.go:462]: This test asserts that errors are propogated when an error occurs during status updates.  I added it to give full coverage to the `ensureSpace` method, which got refactored.

Comment (sadlerap) [reply controllers/spacerequest/spacerequest_controller_test.go]: That said, the name of this test is certainly wrong, which I've fixed in 50e206a.

**Comment by sadlerap** [comment]
> It's always the best practice to create both PRs from branches that have the same name so our e2e logic can pair them and run e2e tests from the e2e PR (instead of master). In this way, you can verify that the changes in e2e tests work as expected with the changes from the operator and you can merge both PRs together.

I wasn't aware there was automation set up for this case.  In the future, I'll do this, but I'll leave the currently-open PRs for now.

Is there any kind of contribution guide where this is documented?

**Pull request merged**
